### PR TITLE
Which do I use? `attributedText` or `textStorage`?

### PR DESCRIPTION
**Pull request by rjchatfield**
Which do I use? `attributedText` or `textStorage`? Because they aren't the same!

> "_TextStorage is source of truth_" 

... yet this change fails a test.

| Expected | Failure |
|---|---|
| reference_1_DCDC6B2B-08DB-4542-9CFB-FE6BBE1B9F8A | failure_2_DCDC6B2B-08DB-4542-9CFB-FE6BBE1B9F8A |

---

The reason I went down this path...

We have been seeing a lot of out of bounds crashes (I'm not able to reproduce them, sadly). We started adding bounds checks in various places. And then I remembered 

> "_Oh, we have our own `TextStorage` class. I should be able add the guards there once._" 

So I added the guards. Here's an example for `attributedSubstring(range:)`:

```objc
- (NSAttributedString *)attributedSubstringFromRange:(NSRange)range {
    if ([self boundsCheck:range]) {
        return [_storage attributedSubstringFromRange: range];
    } else {
        return [[NSAttributedString alloc] init];
    }
}
```

Tests were still passing. I set some break points and realised we weren't always calling those places. I noticed that some places we read or write from either `attributedText` or `textStorage`. Then I thought...

> "_Yay! This was probably the reason for the out of bounds. Now we have a single source of truth._"

So I ran the tests... and they failed. I closed my laptop and cried myself to sleep. I would very much like to understand which one I'm meant to use and when not to use.

### CODE REVIEW

**Comment by rjchatfield** [comment]
To clarify, this line was the only thing that broke a test. 

If you like I could raise a PR with every _other_ usage of `attributedText` switched to `textStorage` and you'd have passing tests. I'd certainly like to try it in prod to see if I have less out of bounds crashes. 

Separately, I could add all the bounds checks in `TextStorage` to see if it helps.

But my purpose for this PR is to start the investigation into "what is the difference?" and "which should I be using?"

**Comment by rjchatfield** [comment]
Today I noticed this: 

image

We have a call to `EditorView.attributedText` which is forwarded to `richTextView.attributedText`.

```swift
/// Text to be set in the `EditorView`
public var attributedText: NSAttributedString {
    get { richTextView.attributedText }
    set { richTextView.attributedText = newValue }
}
```

But the stack trace for `-[UITextView attributedText]` eventually calls `-[NSAttributedString attributedSubstringFromRange:]` which eventually calls into our `-[PRTextStorage attributesAtIndex:effectiveRange:]`.

So `attributedText` makes a copy of `PRTextStorage`, bridging everything from Objective-C to Swift. Which seems wasteful, particularly in this example where we're just wanting `contentLength`.

---

So now I have two primary concerns here:
- Expensive copies of TextStorage (especially when needing a content length)
- Sometimes incorrect bounds (because I've seen the ranges can differ)

Perhaps its worth experimenting with using TextStorage directly in more places. But the question still remains: "When should I use one and when to use the other?"

**Comment by rjchatfield** [comment]
> TL;DR in extreme cases, the benefit is extreme. But in normal cases it's less significant.

We have a particularly brutal test case for the Renderer (far beyond our normal content size or complexity). I did a quick profile of `master` to see how long it takes to layout.

image

Blocks main for about 101 seconds. :omgcat:

image

---

The Test: I replaced all the calls of `attributedText` with `textStorage` in Proton and re-ran the same example, and it is much faster.

image

Blocks main for 30 seconds (3x faster)

image

I have a much smaller test example that is still quite large (probably the upper limit of our customer use case). Looking in Instruments at the first paint I can see time spent in the TableViewCell's `systemLayoutSizeFitting(...)` went from 283ms to 240ms. This is so slow because we call `layoutIfNeeded()` (which solves an unrelated problem for us). But still 16% faster for free?

---

Worth noting that the Tests still pass. :) 

I'll continue to investigate and see if there's any particular hot paths here.

**Comment by rjchatfield** [comment]
Instead of profiling the CPU, this might be effecting the memory footprint more. This is the Allocations instrument:

## All Heap Allocations

|| Total Bytes | # Total |
|---|---|---|
| `master` | 10.83 GB | 16,260,170 |
| Stop calling `attributedString` | 1.89 GB | 12,490,447 |
| **Improvement** | -82% | -23% |

image

**Comment by vox-humana** [comment]
What a fascinating reading 👍 I like that you brought this question. I think you are right that every access to `attributedText` actually creates copy of the whole storage. That's why we probably need to use storage directly. However, personally I don't have enough experience with TextKit for being 100% sure. Also, Apple docs don't help as usual. Let's see how far we can go.

**Comment by rjchatfield** [comment]
I'm closing this PR
1. This is not a "pull request", it's just an investigation 
2. A small amount of change has been made in another PR to address excess allocations #93
